### PR TITLE
SDK-1884: Cypress SDK not wrapping A11Y commands appropriately

### PR DESCRIPTION
**Pull request by osho-20**
- Fixed Cypress A11y Scan
- Added launch Options

### CODE REVIEW

File: bin/accessibility-automation/cypress/index.js [hunk 1]
```diff
@@ -6,6 +6,15 @@ const browserStackLog = (message) => {
   }
   
 const commandsToWrap = ['visit', 'click', 'type', 'request', 'dblclick', 'rightclick', 'clear', 'check', 'uncheck', 'select', 'trigger', 'selectFile', 'scrollIntoView', 'scroll', 'scrollTo', 'blur', 'focus', 'go', 'reload', 'submit', 'viewport', 'origin'];
+const commandToOverwrite = ['visit', 'click', 'type', 'request', 'dblclick', 'rightclick', 'clear', 'check', 'uncheck', 'select', 'trigger', 'selectFile', 'scrollIntoView', 'scrollTo', 'blur', 'focus', 'go', 'reload', 'submit', 'viewport', 'origin'];
```
Comment (sauravdas1997) [inline bin/accessibility-automation/cypress/index.js:9]: any reason for removing `scroll` from the list ?

File: bin/accessibility-automation/cypress/index.js [hunk 2]
```diff
@@ -18,7 +27,7 @@ new Promise(async (resolve, reject) => {
         return win.document.querySelector("#accessibility-automation-element");
     }
 
-    function waitForScannerReadiness(retryCount = 30, retryInterval = 100) {
+    function waitForScannerReadiness(retryCount = 100, retryInterval = 100) {
```
Comment (sauravdas1997) [inline bin/accessibility-automation/cypress/index.js:30]: Is `retryCount` number intentional ? Why 100 ?

File: bin/accessibility-automation/cypress/index.js (hunk 1 above)
Comment (osho-20) [reply bin/accessibility-automation/cypress/index.js:9]: scroll is not a default cypress version. This error was occuring.

File: bin/accessibility-automation/cypress/index.js [hunk 3]
```diff
@@ -337,9 +329,11 @@ Cypress.Commands.add('performScan', () => {
         }
         cy.window().then(async (win) => {
             browserStackLog(`Performing accessibility scan`);
-            await performScan(win);
+            cy.wrap(performScan(win), {timeout:40000});
```
Comment (sauravdas1997) [inline bin/accessibility-automation/cypress/index.js:332]: This sets a fixed timeout, which can break the test if mochaTimeout is less than 40s

Comment (osho-20) [reply bin/accessibility-automation/cypress/index.js:332]: if the response from performScan take less time no timeout will occur. Eventually if response does not come and mocha timeout occurs it should be considered as timeout.

File: bin/helpers/utils.js [hunk 4]
```diff
@@ -1775,3 +1775,24 @@ exports.getMajorVersion = (version) => {
...
+
+exports.decodeJWTToken = (token) => {
+  try{
+    const parts = token.split('.');
+    if (parts.length < 2) {
+      throw new Error('Invalid JWT token');
+    }
+    const header = JSON.parse(base64UrlDecode(parts[0]));
+    const payload = JSON.parse(base64UrlDecode(parts[1]));
+    return { header, payload };
```
Comment (sauravdas1997) [inline bin/helpers/utils.js:1793]: why to parse header if we are not using

Comment (sauravdas1997) [reply bin/helpers/utils.js:1793]: also let's not return it if not used

Comment (osho-20) [reply bin/helpers/utils.js:1793]: i made it a generic function to get used in future if need headers

File: bin/helpers/utils.js [hunk 5]
```diff
@@ -1775,3 +1775,24 @@ exports.getMajorVersion = (version) => {
...
+  } catch (error) {
+    logger.err(error.message);
```
Comment (amaanbs) [inline bin/helpers/utils.js:1794]: Let's add some descriptor text to this error log.

File: bin/accessibility-automation/cypress/index.js [hunk 6]
```diff
@@ -29,9 +69,11 @@ new Promise(async (resolve, reject) => {
                     "Accessibility Automation Scanner is not ready on the page."
                 )
                 );
+				return;
```
Comment (07souravkunda) [inline bin/accessibility-automation/cypress/index.js:72]: shouldn't promise be resolved?

File: bin/accessibility-automation/cypress/index.js [hunk 7]
```diff
@@ -43,6 +85,7 @@ new Promise(async (resolve, reject) => {
         function onScanComplete() {
             win.removeEventListener("A11Y_SCAN_FINISHED", onScanComplete);
             resolve();
+			return;
```
Comment (07souravkunda) [inline bin/accessibility-automation/cypress/index.js:88]: shouldn't promise be resolved?

**Comment by osho-20** [comment]
RUN_SCA

File: bin/accessibility-automation/cypress/index.js [hunk 8]
```diff
@@ -111,7 +159,8 @@ new Promise((resolve) => {
             .then(getSummary)
             .catch((err) => {
             resolve();
-            });
+			return;
```
Comment (07souravkunda) [inline bin/accessibility-automation/cypress/index.js:162]: resolving do we need return?

File: bin/accessibility-automation/plugin/index.js [hunk 9]
```diff
@@ -30,7 +32,16 @@ const browserstackAccessibility = (on, config) => {
         }
         if (browser_validation) {
           const ally_path = path.dirname(process.env.ACCESSIBILITY_EXTENSION_PATH)
+          const {_, payload} = decodeJWTToken(process.env.ACCESSIBILITY_AUTH);
           launchOptions.extensions.push(ally_path);
+          if(!utils.isUndefined(payload) && !utils.isUndefined(payload.a11y_core_config) && payload.a11y_core_config.domForge === true) {
```
Comment (07souravkunda) [inline bin/accessibility-automation/plugin/index.js:37]: significance of these?

File: bin/accessibility-automation/cypress/index.js [hunk 10]
```diff
@@ -6,19 +6,59 @@ const browserStackLog = (message) => {
   }
   
 const commandsToWrap = ['visit', 'click', 'type', 'request', 'dblclick', 'rightclick', 'clear', 'check', 'uncheck', 'select', 'trigger', 'selectFile', 'scrollIntoView', 'scroll', 'scrollTo', 'blur', 'focus', 'go', 'reload', 'submit', 'viewport', 'origin'];
+const commandToOverwrite = ['visit', 'click', 'type', 'request', 'dblclick', 'rightclick', 'clear', 'check', 'uncheck', 'select', 'trigger', 'selectFile', 'scrollIntoView', 'scrollTo', 'blur', 'focus', 'go', 'reload', 'submit', 'viewport', 'origin'];
+const performModifiedScan = (originalFn, Subject, stateType, ...args) => {
```
Comment (07souravkunda) [inline bin/accessibility-automation/cypress/index.js:10]: please add some comments explaining the below function for future ref

File: bin/accessibility-automation/cypress/index.js [hunk 11]
```diff
@@ -6,19 +6,59 @@ const browserStackLog = (message) => {
...
+            else {
+                setTimeout = null;
+            }
+            if((orgSC1 = (orgSC2 = cy).subjectChain) !== null && orgSC1 !== void 0){
+                cypressCommandChain = orgSC1.call(orgSC2);
+            }
+            else {
+                cypressCommandChain = null;
+            }
+            customChaining.performScanSubjectQuery(cypressCommandChain, setTimeout).then({timeout: 30000}, (newSubject) => originalFn(...changeSub(args, stateType, newSubject)));
```
Comment (07souravkunda) [inline bin/accessibility-automation/cypress/index.js:43]:
```suggestion
 function runCutomizedCommand() {
    if (!Subject) {
        let cypressCommandSubject = (cy.subject?.call(cy)) ?? null;
        customChaining.then(() => cypressCommandSubject).then(() => { originalFn(...args); });
    } else {
        let setTimeout = args.find(arg => arg?.timeout)?.timeout ?? null;
        let cypressCommandChain = (cy.subjectChain?.call(cy)) ?? null;
        customChaining.performScanSubjectQuery(cypressCommandChain, setTimeout).then({ timeout: 30000 }, newSubject => originalFn(...changeSub(args, stateType, newSubject)));
    }
}
```

**Review by 07souravkunda — CHANGES_REQUESTED** [review]

File: bin/accessibility-automation/cypress/index.js (hunk 8 above)
Comment (osho-20) [reply bin/accessibility-automation/cypress/index.js:162]: no

File: bin/accessibility-automation/plugin/index.js (hunk 9 above)
Comment (osho-20) [reply bin/accessibility-automation/plugin/index.js:37]: other ticket SDK-1907

File: bin/accessibility-automation/cypress/index.js [hunk 12]
```diff
@@ -6,19 +6,41 @@ const browserStackLog = (message) => {
...
+const commandToOverwrite = ['visit', 'click', 'type', 'request', 'dblclick', 'rightclick', 'clear', 'check', 'uncheck', 'select', 'trigger', 'selectFile', 'scrollIntoView', 'scrollTo', 'blur', 'focus', 'go', 'reload', 'submit', 'viewport', 'origin'];
+const performModifiedScan = (originalFn, Subject, stateType, ...args) => {
+    let customChaining = cy.wrap(null).performScan();
+    function changeSub(args, stateType, newSubject) {
+        if (stateType !== 'parent') {
+            return [newSubject, ...args.slice(1)];
+        }
+        return args;
+    }
+    function runCutomizedCommand() {
```
Comment (amaanbs) [inline bin/accessibility-automation/cypress/index.js:18]: Use arrow functions

File: bin/accessibility-automation/cypress/index.js [hunk 13]
```diff
@@ -6,19 +6,41 @@ const browserStackLog = (message) => {
   }
   
 const commandsToWrap = ['visit', 'click', 'type', 'request', 'dblclick', 'rightclick', 'clear', 'check', 'uncheck', 'select', 'trigger', 'selectFile', 'scrollIntoView', 'scroll', 'scrollTo', 'blur', 'focus', 'go', 'reload', 'submit', 'viewport', 'origin'];
+const commandToOverwrite = ['visit', 'click', 'type', 'request', 'dblclick', 'rightclick', 'clear', 'check', 'uncheck', 'select', 'trigger', 'selectFile', 'scrollIntoView', 'scrollTo', 'blur', 'focus', 'go', 'reload', 'submit', 'viewport', 'origin'];
+const performModifiedScan = (originalFn, Subject, stateType, ...args) => {
+    let customChaining = cy.wrap(null).performScan();
+    function changeSub(args, stateType, newSubject) {
```
Comment (amaanbs) [inline bin/accessibility-automation/cypress/index.js:12]: Use arrow functions

File: bin/accessibility-automation/cypress/index.js [hunk 14]
```diff
@@ -6,19 +6,41 @@ const browserStackLog = (message) => {
...
+        return args;
+    }
+    function runCutomizedCommand() {
+        if (!Subject) {
+            let cypressCommandSubject = (cy.subject?.call(cy)) ?? null;
+            customChaining.then(() => cypressCommandSubject).then(() => { originalFn(...args); });
+        } else {
+            let setTimeout = args.find(arg => arg?.timeout)?.timeout ?? null;
+            let cypressCommandChain = (cy.subjectChain?.call(cy)) ?? null;
+            customChaining.performScanSubjectQuery(cypressCommandChain, setTimeout).then({ timeout: 30000 }, newSubject => originalFn(...changeSub(args, stateType, newSubject)));
```
Comment (amaanbs) [inline bin/accessibility-automation/cypress/index.js:25]: Change to 10 seconds.

**Review by amaanbs — CHANGES_REQUESTED** [review]

**Review by amaanbs — APPROVED** [review]

**Review by amaanbs — DISMISSED** [review]

**Review by 07souravkunda — APPROVED** [review]

**Review by asambstack — APPROVED** [review]

**Review by pranavj1001 — APPROVED** [review]
Approved for base branch change

**Pull request merged**